helix-term/src/ui/picker/handlers.rs

```diff
@@ -0,0 +1,168 @@
+            let text = doc.text().clone();
+            let loader = editor.syn_loader.clone();
+
+            tokio::task::spawn_blocking(move || {
+                let Some(syntax) =
+                    language_config
+                        .highlight_config(&loader.load().scopes())
+                        .and_then(|highlight_config| {
+                            helix_core::Syntax::new(
+                                text.slice(..),
+                                highlight_config,
+                                loader,
+                            )
+                        }) else {
+                    log::info!("highlighting picker item failed");
+                    return;
+                };
+
+                crate::job::dispatch_blocking(move |editor, compositor| {
+                    let Some(Overlay { content: picker, .. }) = compositor.find::<Overlay<Picker<T, D>>>() else {
+                        log::info!("picker closed before syntax highlighting finished");
+                        return;
+                    };
+                    let Some(CachedPreview::Document(ref mut doc)) = picker.preview_cache.get_mut(&path) else {
+                        return
+                    };
+                    let diagnostics = helix_view::Editor::doc_diagnostics(
+                        &editor.language_servers,
+                        &editor.diagnostics,
+                        doc,
+                    );
+                    doc.replace_diagnostics(diagnostics, &[], None);
+                    doc.syntax = Some(syntax);
+                });
+            });
+        });
+    }
+}
+
+pub(super) struct DynamicQueryHandler<T: 'static + Send + Sync, D: 'static + Send + Sync> {
+    callback: Arc<DynQueryCallback<T, D>>,
+    last_query: Arc<str>,
+    query: Option<Arc<str>>,
+}
+
+impl<T: 'static + Send + Sync, D: 'static + Send + Sync> DynamicQueryHandler<T, D> {
+    pub(super) fn new(callback: DynQueryCallback<T, D>) -> Self {
+        Self {
+            callback: Arc::new(callback),
+            last_query: "".into(),
+            query: None,
+        }
+    }
+}
+
+impl<T: 'static + Send + Sync, D: 'static + Send + Sync> AsyncHook for DynamicQueryHandler<T, D> {
+    type Event = Arc<str>;
+
+    fn handle_event(&mut self, query: Self::Event, _timeout: Option<Instant>) -> Option<Instant> {
+        if query == self.last_query {
+            // If the search query reverts to the last one we requested, no need to
+            // make a new request.
+            self.query = None;
+            None
+        } else {
+            self.query = Some(query);
+            Some(Instant::now() + Duration::from_millis(275))
+        }
+    }
+
+    fn finish_debounce(&mut self) {
+        let Some(query) = self.query.take() else { return };
+        self.last_query = query.clone();
+        let callback = self.callback.clone();
+
+        crate::job::dispatch_blocking(move |editor, compositor| {
```

Comment:
nit I think importing job so that its `job::dispatch_blocking` is slightly easier to read